### PR TITLE
Hit Fused Addmm Path for Nd tensors

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106740
* __->__ #106433
* #106739

Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/39661 https://github.com/pytorch/pytorch/pull/74070

Seems this has been done a few times in different contexts :)

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -75,19 +79,20 @@ Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Ten
...
   }
 #endif
-  if (input.dim() == 2 && bias->defined()) {
+  const auto input_dim = input.dim();
+  if (input_dim == 2 && bias->defined()) {
     // Fused op is marginally faster.
     return at::addmm(*bias, input, weight.t());
   }
-  if (input.dim() == 3 && bias->defined() && !input.is_xla()) {
+  if ((input_dim == 3 || input_dim == 4) && bias->defined() && !input.is_xla()) {
```
Comment (vadimkantorov) [inline aten/src/ATen/native/Linear.cpp:87]: what is the motivation behind only fusing for 3-dim and 4-dim inputs? any contig inputs should work fine, right? (as can reshape without any perf hit)

Comment (eellison) [reply aten/src/ATen/native/Linear.cpp:87]: yea, no motivation, just didnt have real world models to test on. I can update to hit it for all dimensions.

Comment (vadimkantorov) [reply aten/src/ATen/native/Linear.cpp:87]: in my personal view, the less hidden differences/surprises like this, the better...

**Review by eellison — COMMENTED** [review]
cc @alexmsettle this is breaking `test_inductor_sequence_nr`. would you take a look ?

File: test/dynamo/test_aot_autograd.py [hunk 2]
```diff
@@ -810,39 +810,40 @@ def _prepare_model_args():
...
+7|aten.sum.dim_IntList|
+7|aten.view.default|
+6|aten.t.default|
+5|aten.view.default|
+4|aten.view.default|
+2|aten.threshold_backward.default|
+1|aten.native_batch_norm_backward.default|
+0|aten.convolution_backward.default|
+"""
             ),
```
Comment (alexmsettle) [inline test/dynamo/test_aot_autograd.py:847]: @eellison - this change didn't fix the failing test?  It looks like your change set changed the aten ops which would change the fx listing of aten ops.  I would have recommended just changing the expected output like you did here.  Let me know if you need me to dig into this further.

**Pull request closed without merging**